zmq.go

```diff
@@ -396,36 +405,53 @@ func Subscribe(addr string, topics []string, timeout time.Duration) (*Conn, erro
 }
 
 // Receive a message from the publisher. It blocks until a new message is
-// received.
+// received. If the connection times out and it was not explicitly terminated,
+// then a timeout error is returned. Otherwise, if it was explicitly terminated,
+// then io.EOF is returned.
 func (c *Conn) Receive() ([][]byte, error) {
 	messages, err := c.readMessage()
 	// If the error is either nil or a non-EOF error, we return it as-is.
 	if err != io.EOF {
 		return messages, err
 	}
-	// We got an EOF, so our socket is disconnected. We attempt to
-	// reconnect. If successful, replace the existing connection with the
-	// new one. Either way, return a timeout error.
+
+	// We got an EOF, so our socket is disconnected. If the connection was
+	// explicitly terminated, we'll return the EOF error.
+	select {
+	case <-c.quit:
+		return nil, io.EOF
+	default:
+	}
+
+	// Otherwise, we'll attempt to reconnect. If successful, we'll replace
+	// the existing connection with the new one. Either way, return a
+	// timeout error.
 	errTimeout := &net.OpError{
 		Op:     "read",
 		Net:    c.conn.LocalAddr().Network(),
 		Source: c.conn.LocalAddr(),
 		Addr:   c.conn.RemoteAddr(),
 		Err:    &reconnectError{err},
 	}
-	newConn, err := Subscribe(c.conn.RemoteAddr().String(), c.topics,
-		c.timeout)
+	newConn, err := Subscribe(
+		c.conn.RemoteAddr().String(), c.topics, c.timeout,
+	)
 	if err != nil {
 		// Prevent CPU overuse by refused reconnection attempts.
 		time.Sleep(c.timeout)
 	} else {
-		c.Close()
-		*c = *newConn
+		c.conn.Close()
```

Comment:
No error to return here?

Reply:
We still want to return `errTimeout`. Closing this connection seems like more of a best effort thing.